feat: Add openAI compatible APIs to QDrant

varshaprasad96 opened the pull request:
# What does this PR do?
Adds support to Open AI APIs in Qdrant.

<!-- If resolving an issue, uncomment and update the line below -->
 Closes #2463 


## Test Plan
<!-- Describe the tests you ran to verify your changes with result summaries. *Provide clear instructions so the plan can be easily re-executed.* -->

varshaprasad96 commented on llama_stack/providers/inline/vector_io/qdrant/__init__.py:
```diff
@@ -4,14 +4,17 @@
+async def get_provider_impl(config: QdrantVectorIOConfig, deps: dict[Api, Any]):
     from llama_stack.providers.remote.vector_io.qdrant.qdrant import QdrantVectorIOAdapter
 
+    assert isinstance(config, QdrantVectorIOConfig), f"Unexpected config type: {type(config)}"
```
This is just additional validation - can remove it if reviewers feel otherwise.

franciscojavierarceo commented on llama_stack/providers/remote/vector_io/qdrant/qdrant.py:
```diff
@@ -127,21 +123,86 @@ async def delete(self):
+        if not await self.client.collection_exists(metadata_collection):
+            await self.client.create_collection(
+                collection_name=metadata_collection,
+                vectors_config=models.VectorParams(size=1, distance=models.Distance.COSINE),
```
I don't love having the distance metric hard coded instead of being configurable.

franciscojavierarceo reviewed (approved):
small nit i have but this lgtm

hardikjshah commented on llama_stack/providers/inline/vector_io/qdrant/config.py:
```diff
@@ -15,6 +15,7 @@
 @json_schema_type
 class QdrantVectorIOConfig(BaseModel):
     path: str
+    distance_metric: Literal["COSINE", "DOT", "EUCLID", "MANHATTAN"] = "COSINE"
```
This does not fit here. This is the config to setup the provider. 
Not every use case for this provider will use Qdrant with the same distance metric. 

We need this to be at the API layer not the provider configuration layer.

varshaprasad96 replied:
Modified for insert_chunks to accept metadata where distance can be specified.

hardikjshah commented on llama_stack/providers/remote/vector_io/qdrant/qdrant.py:
```diff
@@ -127,28 +129,100 @@ async def delete(self):
+        if not await self.client.collection_exists(metadata_collection):
+            return {}
+
+        # Get all points from metadata collection
+        points = await self.client.scroll(
+            collection_name=metadata_collection,
+            limit=1000,  # Reasonable limit for metadata
+            with_payload=True,
+        )
```
Would be better to search with a filtering condition instead of having an adhoc limit of 1000. 
```
from qdrant_client.models import Filter, FieldCondition, Range

hits = client.search(
   collection_name="my_collection",
   query_vector=query_vector,
   query_filter=Filter(
      must=[  # These conditions are required for search results
            FieldCondition(
               key='rand_number',  # Condition based on values of `rand_number` field.
               range=Range(
                  gte=3  # Select only those results where `rand_number` >= 3
               )
            )
      ]
   ),
   limit=5  # Return 5 closest points
)
``` 
https://python-client.qdrant.tech/

varshaprasad96 replied:
@hardikjshah Looking at the documentation using scroll seems more appropriate for this specific use case. The `_load_openai_vector_stores method` is designed to load all vector store metadata during initialization, not perform similarity search. Using search would require a query vector (which we don't have for metadata retrieval) and would be inefficient for bulk data operations. According to the Qdrant documentation, scroll is specifically designed for bulk data retrieval and iteration through collections, while search is optimized for similarity queries with filters. 

The 1000 limit is considering that vector store metadata is small and most deployments won't exceed this number. The other option is to implement scroll with pagination if that fits better (https://qdrant.github.io/qdrant/redoc/index.html#tag/Points/operation/batch_update)? Wdyt?

derekhiggins commented on llama_stack/providers/remote/vector_io/qdrant/qdrant.py:
```diff
@@ -127,28 +129,100 @@ async def delete(self):
+        self,
+        config: RemoteQdrantVectorIOConfig | InlineQdrantVectorIOConfig,
+        inference_api: Api.inference,
+        files_api: Files | None,
```
files_api needs to also be passed into remote provider

```
  File "/home/derekh/workarea/llama-stack/llama_stack/providers/remote/vector_io/qdrant/__init__.py", line 15, in get_adapter_impl
    impl = QdrantVectorIOAdapter(config, deps[Api.inference])
TypeError: QdrantVectorIOAdapter.__init__() missing 1 required positional argument: 'files_api'

```

derekhiggins commented on llama_stack/providers/remote/vector_io/qdrant/qdrant.py:
```diff
@@ -127,28 +129,100 @@ async def delete(self):
+        self.openai_vector_stores: dict[str, dict[str, Any]] = {}
 
     async def initialize(self) -> None:
         self.client = AsyncQdrantClient(**self.config.model_dump(exclude_none=True))
```
Hitting an error here in the remote use case

```
  File "/home/derekh/workarea/llama-stack/llama_stack/providers/remote/vector_io/qdrant/__init__.py", line 17, in get_adapter_impl
    await impl.initialize()
  File "/home/derekh/workarea/llama-stack/llama_stack/providers/utils/telemetry/trace_protocol.py", line 103, in async_wrapper
    result = await method(self, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/derekh/workarea/llama-stack/llama_stack/providers/remote/vector_io/qdrant/qdrant.py", line 148, in initialize
    self.client = AsyncQdrantClient(**self.config.model_dump(exclude_none=True))
                  ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/derekh/workarea/llama-stack/venv/lib/python3.13/site-packages/qdrant_client/async_qdrant_client.py", line 126, in __init__
    self._client = AsyncQdrantRemote(
                   ~~~~~~~~~~~~~~~~~^
        url=url,
        ^^^^^^^^
    ...<11 lines>...
        **kwargs,
        ^^^^^^^^^
    )
    ^
  File "/home/derekh/workarea/llama-stack/venv/lib/python3.13/site-packages/qdrant_client/async_qdrant_remote.py", line 165, in __init__
    self.openapi_client: AsyncApis[AsyncApiClient] = AsyncApis(
                                                     ~~~~~~~~~^
        host=self.rest_uri, **self._rest_args
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/derekh/workarea/llama-stack/venv/lib/python3.13/site-packages/qdrant_client/http/api_client.py", line 26, in __init__
    self.client = AsyncApiClient(host, **kwargs)
                  ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^
  File "/home/derekh/workarea/llama-stack/venv/lib/python3.13/site-packages/qdrant_client/http/api_client.py", line 158, in __init__
    self._async_client = AsyncClient(**kwargs)
                         ~~~~~~~~~~~^^^^^^^^^^
TypeError: AsyncClient.__init__() got an unexpected keyword argument 'distance_metric'

```

varshaprasad96 replied:
@derekhiggins I've removed distance_metric from config and reverted to back to what it was. It should work now.

varshaprasad96 commented:
The PR is still in progress - fixing a few things based on discussions on Discord

varshaprasad96 commented:
The failure in here seems to be unrelated to qdrant changes:
```
__________________ test_responses_store[txt=3B-tools1-False] ___________________
tests/integration/agents/test_openai_responses.py:52: in test_responses_store
    response = client.responses.create(
.venv/lib/python3.13/site-packages/openai/_utils/_utils.py:279: in wrapper
    return func(*args, **kwargs)
.venv/lib/python3.13/site-packages/openai/resources/responses/responses.py:603: in create
    return self._post(
.venv/lib/python3.13/site-packages/openai/_base_client.py:1242: in post
    return cast(ResponseT, self.request(cast_to, opts, stream=stream, stream_cls=stream_cls))
.venv/lib/python3.13/site-packages/openai/_base_client.py:919: in request
    return self._request(
.venv/lib/python3.13/site-packages/openai/_base_client.py:1023: in _request
    raise self._make_status_error_from_response(err.response) from None
E   openai.BadRequestError: Error code: 400 - {'detail': 'Invalid value: Unknown tool get_weather  called'}
```

varshaprasad96 commented on llama_stack/providers/inline/vector_io/sqlite_vec/sqlite_vec.py:
```diff
@@ -729,7 +731,13 @@ def _delete():
+        vector_db_id: str,
+        chunks: list[Chunk],
+        ttl_seconds: int | None = None,
+        params: dict[str, Any] | None = None,
```
Adding distance_metric in params instead of mentioning it explicitly - as not all providers need it. We can later expand params to include any other knobs for providers as they see fit.

varshaprasad96 commented:
@franciscojavierarceo @hardikjshah @derekhiggins - I was able to test these changes with remote Qdrant using the test script in [here](https://github.com/meta-llama/llama-stack/blob/aaa8c7b724b63815da3680604a330bb955ebee27/test_qdrant_apis.py). It works as expected:
```
Testing Qdrant Vector IO APIs...
Server URL: http://localhost:8321
--------------------------------------------------

1. Testing health check...
Health check passed

2. Testing list vector DBs...
List vector DBs successful: 2 DBs found
   - test-qdrant-db
   - test-store

3. Testing register vector DB...
Register vector DB successful

4. Testing list vector DBs (after registration)...
List vector DBs successful: 2 DBs found
   - test-qdrant-db
   - test-store

5. Testing OpenAI vector store APIs...
Create vector store successful: test-store
List vector stores successful: 1 stores found
```

hardikjshah commented on llama_stack/providers/remote/vector_io/qdrant/qdrant.py:
```diff
@@ -41,6 +35,13 @@
 CHUNK_ID_KEY = "_chunk_id"
 
+# KV store prefixes for OpenAI vector stores
+OPENAI_VECTOR_STORES_PREFIX = "openai_vector_stores:"
```
Similar to https://github.com/meta-llama/llama-stack/pull/2470
Lets make these provider specific,
```
OPENAI_VECTOR_STORES_PREFIX = f"openai_vector_stores:qdrant:{VERSION}::"
```

hardikjshah commented on llama_stack/providers/remote/vector_io/qdrant/qdrant.py:
```diff
@@ -41,6 +35,13 @@
+OPENAI_VECTOR_STORES_FILES_PREFIX = "openai_vector_stores_files:"
+OPENAI_VECTOR_STORES_FILES_CONTENTS_PREFIX = "openai_vector_stores_files_contents:"
+
+VECTOR_DBS_PREFIX = "vector_dbs:"
```
same here -- make this `qdrant`  specific.

hardikjshah reviewed (changes requested):
suggested couple of small changes

varshaprasad96 commented:
@franciscojavierarceo @hardikjshah addressed reviews, the PR should be good to go. PTAL.

franciscojavierarceo reviewed (commented):
can you enable Qdrant in CI?

varshaprasad96 commented:
> can you enable Qdrant in CI?

Yes, I have a follow up PR to do so. Just to make sure that the feature implementation gets in first.

varshaprasad96 commented:
> it strikes me as something we should avoid. it represents an abstraction leak. a user sends params when they know details of the backend provider implementation. this will tie user code to a specific provider, and likely a specific version of a provider.

@mattf the intention to not explicitly provide it as a typed parameter in the function and rather make it pass as a metadata was to allow flexibility to let providers validate and accept what is relevant to them. Distance metric may be relevant to Qdrant implementation but not be so for others. 

> it would make more sense to include the specific params during db construction, where the user has to make decisions about the provider and its configuration.

> also, i expect the distance metric can be derived from information about the model.

This is what was implemented in the initial commit: ref: https://github.com/meta-llama/llama-stack/pull/2465#discussion_r2155577091 and it was suggested that this need not explicitly be configured with the DB, and should be allowed to be modified.

franciscojavierarceo commented:
>But even then, I’m not convinced it’s necessary. Embedding models are typically designed with a specific distance metric in mind—most often cosine similarity or dot product. Using a different metric than what the model was trained for can lead to degraded performance or semantic mismatches. This suggests that the distance metric should either be inferred from the model or defined at model creation time—not passed at insert time.

I generally agree here and it's reasonably pragmatic but in my opinion distance metrics should be defined during retrieval or on configuration of the vector store. I think the only reason for configuring the distance metric at the configuration of the provider is because of laziness of an end user, even though it's, of course, technically better to pass this during a request.  I think @hardikjshah wasn't preferential towards this though. 

I don't think it makes sense to pass it during ingestion as I don't see any obvious benefit to it.

>@franciscojavierarceo would adding the distance metric to the model or the vector DB satisfy your concern about hardcoding?

Personally, I'm not overly fond of this option as the mapping between distance metric and provider isn't one-to-one.

These are, of course, just my opinions though.

franciscojavierarceo commented:
@varshaprasad96 can you update `tests/unit/providers/vector_io/conftest.py` to include inline qdrant?

franciscojavierarceo reviewed (changes requested):
let's incorporate feedback from @mattf and @hardikjshah 

thanks for working through this @varshaprasad96 !!

varshaprasad96 commented:
/hold - trying to fix `conftest.py` for qdrant

varshaprasad96 commented:
@mattf @hardikjshah @franciscojavierarceo - I've refactored the PR, and removed the use of distance_metrics. Have it to "cosine" as it was before: https://github.com/meta-llama/llama-stack/blob/51d9fd48083de32e9796f00030b57be10a8ac7bd/llama_stack/providers/remote/vector_io/qdrant/qdrant.py#L69

Based on previous discussions, it probably makes sense to pass distance_metric along with [Vector_DB](https://github.com/meta-llama/llama-stack/blob/51d9fd48083de32e9796f00030b57be10a8ac7bd/llama_stack/apis/vector_dbs/vector_dbs.py#L17) where we also have model specific embedding dimension. Distance metric is also required by [milvus](https://github.com/meta-llama/llama-stack/blob/51d9fd48083de32e9796f00030b57be10a8ac7bd/llama_stack/providers/remote/vector_io/milvus/milvus.py#L115) - and this can refactored in a separate PR.

ehhuang commented on llama_stack/providers/remote/vector_io/qdrant/qdrant.py:
```diff
@@ -132,25 +139,139 @@ async def delete(self):
     async def shutdown(self) -> None:
         await self.client.close()
 
+    # OpenAI Vector Store Mixin abstract method implementations
```
are the below just duplicated? why?

ehhuang commented:
Could you please update .github/workflows/integration-vector-io-tests.yml so that the tests are run in CI?

ehhuang commented:
> @ehhuang @franciscojavierarceo The PR has been rebased with main, and also tested with the following script. It is good to go:
> 
> ```
> import asyncio
> import io
> import json
> import tempfile
> import time
> from pathlib import Path
> from typing import Any, Dict
> 
> from llama_stack_client import LlamaStackClient
> 
> 
> class QdrantVectorStoreAPITester:
>     """Test class for Qdrant OpenAI Vector Store APIs using Llama Stack client."""
>     
>     def __init__(self, base_url: str = "http://localhost:8321/"):
>         self.base_url = base_url
>         self.client = LlamaStackClient(base_url=base_url)
>         self.test_files = []
>         self.test_vector_stores = []
>     
>     async def __aenter__(self):
>         return self
>     
>     async def __aexit__(self, exc_type, exc_val, exc_tb):
>         await self.cleanup()
>     
>     async def cleanup(self):
>         """Clean up test resources."""
>         print("\nCleaning up test resources...")
>         
>         # Delete vector stores
>         for vector_store_id in self.test_vector_stores:
>             try:
>                 self.client.vector_stores.delete(vector_store_id)
>                 print(f"    Vector store {vector_store_id} deleted")
>             except Exception as e:
>                 print(f"    Failed to delete vector store {vector_store_id}: {e}")
>         
>         # Delete files
>         for file_id in self.test_files:
>             try:
>                 self.client.files.delete(file_id)
>                 print(f"    File {file_id} deleted")
>             except Exception as e:
>                 print(f"    Failed to delete file {file_id}: {e}")
>     
>     def create_test_file(self, content: str, filename: str = "test.txt") -> io.BytesIO:
>         """Create a test file in memory."""
>         file_content = content.encode('utf-8')
>         file_buffer = io.BytesIO(file_content)
>         file_buffer.name = filename
>         return file_buffer
>     
>     async def test_files_api(self):
>         """Test Files API using client methods."""
>         print("\nTesting Files API...")
>         
>         # Upload a single file for testing
>         test_content = "Artificial intelligence (AI) is intelligence demonstrated by machines."
>         test_file = self.create_test_file(test_content, "ai_document.txt")
>         
>         # Upload file using client with raw response for debugging
>         print("    Sending file upload request with raw response...")
>         
>         raw_response = self.client.files.with_raw_response.create(
>             file=test_file,
>             purpose="assistants"
>         )
>         
>         # Print response details for debugging
>         print(f"    File upload response status: {raw_response.status_code}")
>         print(f"    File upload response body: {raw_response.text()}")
>         
>         # Parse the response
>         file_info = raw_response.parse()
>         
>         file_id1 = file_info.id
>         self.test_files.append(file_id1)
>         print(f"    File uploaded successfully: {file_id1}")
>         
>         # Test list files
>         files_list = self.client.files.list()
>         file_ids = [f.id for f in files_list.data]
>         assert file_id1 in file_ids, "Uploaded file not found in list"
>         print(f"    File listing successful: {len(files_list.data)} files found")
>         
>         # Test retrieve file
>         retrieved_file = self.client.files.retrieve(file_id1)
>         assert retrieved_file.id == file_id1
>         # Note: filename might be auto-generated, so we just check that it exists
>         assert retrieved_file.filename is not None
>         print(f"    File retrieval successful (filename: {retrieved_file.filename})")
>         
>         # Test retrieve file content
>         file_content = self.client.files.content(file_id1)
>         # File content might already be a string
>         if isinstance(file_content, bytes):
>             content_text = file_content.decode('utf-8')
>         else:
>             content_text = file_content
>         assert content_text == test_content, "File content doesn't match"
>         print(f"    File content retrieval successful")
>         
>         return file_id1, None
>     
>     async def test_vector_store_api(self, file_ids: list[str]):
>         """Test Vector Store API using client methods."""
>         print("\nTesting Vector Store API...")
>         
>         # Test 1: Create vector store with raw response for debugging
>         print("  Testing vector store creation...")
>         print("    Sending request with raw response...")
>         
>         raw_response = self.client.vector_stores.with_raw_response.create(
>             name="Test Vector Store",
>             file_ids=file_ids,
>             embedding_model="ollama/nomic-embed-text",
>             embedding_dimension=384,
>             provider_id="qdrant",
>             metadata={"test": True, "description": "Test vector store for API validation"}
>         )
>         
>         # Print response details for debugging
>         print(f"    Response status: {raw_response.status_code}")
>         print(f"    Response body: {raw_response.text()}")
>         
>         # Parse the response
>         create_result = raw_response.parse()
>         
>         vector_store_id = create_result.id
>         self.test_vector_stores.append(vector_store_id)
>         
>         print(f"    Vector store created: {vector_store_id}")
>         assert create_result.name == "Test Vector Store"
>         assert create_result.file_counts.total == len(file_ids)
>         
>         # Wait for all files to be processed (up to 30s)
>         for i in range(30):
>             store = self.client.vector_stores.retrieve(vector_store_id)
>             print(f"    Poll {i+1}: completed={store.file_counts.completed}, failed={store.file_counts.failed}, in_progress={store.file_counts.in_progress}, total={store.file_counts.total}")
>             if store.file_counts.completed == len(file_ids):
>                 break
>             await asyncio.sleep(1)
>         else:
>             final_store = self.client.vector_stores.retrieve(vector_store_id)
>             raise AssertionError(f"Timeout: completed={final_store.file_counts.completed}, failed={final_store.file_counts.failed}, in_progress={final_store.file_counts.in_progress}, total={final_store.file_counts.total}")
>         
>         assert store.file_counts.completed == len(file_ids)
>         
>         # Test 2: List vector stores
>         print("  Testing vector store listing...")
>         stores_list = self.client.vector_stores.list()
>         store_ids = [s.id for s in stores_list.data]
>         assert vector_store_id in store_ids, "Created vector store not found in list"
>         print(f"    Vector store listing successful: {len(stores_list.data)} stores found")
>         
>         # Test 3: Retrieve vector store
>         print("  Testing vector store retrieval...")
>         retrieved_store = self.client.vector_stores.retrieve(vector_store_id)
>         assert retrieved_store.id == vector_store_id
>         assert retrieved_store.name == "Test Vector Store"
>         print(f"    Vector store retrieval successful")
>         
>         # Test 4: Update vector store
>         print("  Testing vector store update...")
>         update_result = self.client.vector_stores.update(
>             vector_store_id,
>             name="Updated Test Vector Store",
>             metadata={"test": True, "description": "Updated test vector store", "updated": True}
>         )
>         
>         assert update_result.name == "Updated Test Vector Store"
>         assert update_result.metadata["updated"] is True
>         print(f"    Vector store update successful")
>         
>         # Test 5: Search vector store
>         print("  Testing vector store search...")
>         search_result = self.client.vector_stores.search(
>             vector_store_id,
>             query="artificial intelligence",
>             max_num_results=5
>         )
>         
>         assert len(search_result.data) > 0
>         print(f"    Vector store search successful: {len(search_result.data)} results found")
>         
>         return vector_store_id
>     
>     async def test_vector_store_files_api(self, vector_store_id: str, file_ids: list[str]):
>         """Test Vector Store Files API using client methods."""
>         print("\nTesting Vector Store Files API...")
>         
>         # Test 1: List files in vector store
>         print("  Testing list files in vector store...")
>         files_list = self.client.vector_stores.files.list(vector_store_id)
>         store_file_ids = [f.id for f in files_list.data]
>         assert all(fid in store_file_ids for fid in file_ids), "Not all files found in vector store"
>         print(f"    List files successful: {len(files_list.data)} files found")
>         
>         # Test 2: Retrieve file from vector store
>         print("  Testing retrieve file from vector store...")
>         file_id = file_ids[0]
>         retrieved_file = self.client.vector_stores.files.retrieve(file_id, vector_store_id=vector_store_id)
>         assert retrieved_file.id == file_id
>         print(f"    Retrieve file successful")
>         
>   
>     
>     async def test_delete_operations(self, vector_store_id: str, file_ids: list[str]):
>         """Test delete operations using client methods."""
>         print("\nTesting delete operations...")
>         
>         # Test 1: Delete file from vector store
>         print("  Testing delete file from vector store...")
>         file_id = file_ids[0]
>         self.client.vector_stores.files.delete(file_id, vector_store_id=vector_store_id)
>         print(f"    Delete file from vector store successful")
>         
>         # Test 2: Delete vector store
>         print("  Testing delete vector store...")
>         self.client.vector_stores.delete(vector_store_id)
>         self.test_vector_stores.remove(vector_store_id)
>         print(f"    Delete vector store successful")
>     
>     async def run_all_tests(self):
>         print("\nStarting Qdrant OpenAI Vector Store API Tests\n" + "="*60 + "\n")
>         file_id1, _ = await self.test_files_api()
>         vector_store_id = await self.test_vector_store_api([file_id1])
>         await self.test_vector_store_files_api(vector_store_id, [file_id1])
>         await self.test_delete_operations(vector_store_id, [file_id1])
> 
> 
> async def main():
>     """Main test function."""
>     async with QdrantVectorStoreAPITester() as tester:
>         await tester.run_all_tests()
>         print("\n" + "="*60)
>         print("All tests completed successfully!")
>         print("Files API: Upload, List, Retrieve, Content")
>         print("Vector Store API: Create, List, Retrieve, Update, Search")
>         print("Vector Store Files API: List, Retrieve, Update, Content")
>         print("Delete Operations: File and Vector Store deletion")
> 
> 
> if __name__ == "__main__":
>     asyncio.run(main())
> ```

I didn't read through this carefully, but does it make sense to incorporate these tests into our existing tests?

franciscojavierarceo commented on llama_stack/providers/remote/vector_io/qdrant/qdrant.py:
```diff
@@ -132,25 +139,103 @@ async def delete(self):
     async def shutdown(self) -> None:
         await self.client.close()
 
+    async def _save_openai_vector_store_file(
```
interesting here you use the kvstore, which is where i think we want to end up but in milvus.py and the other providers we technically implement this natively in the provider for files.

franciscojavierarceo commented on tests/unit/providers/vector_io/conftest.py:
```diff
@@ -236,13 +238,66 @@ async def faiss_vec_adapter(unique_kvstore_config, mock_inference_api, embedding
+
+
+@pytest.fixture
+async def qdrant_vec_adapter(qdrant_vec_db_path, mock_inference_api, embedding_dimension):
```
@ehhuang by creating this adapter and adding it to `vector_io_adapter` on line 293, it enables inline unit tests for qdrant. Additionally, the integration tests are configured for qdrant in `tests/integration/vector_io/test_openai_vector_stores.py`, so the testing should have coverage. 

@varshaprasad96 was there a reason you needed the additional script? I would have just added to those tests.

ehhuang replied:
Shouldn't https://github.com/ehhuang/llama-stack/blob/9736f096f66dd45784537fed6ac49bb900f70f97/.github/workflows/integration-vector-io-tests.yml#L25 be updated to include qdrant?

franciscojavierarceo replied:
I realized it after I left that comment and forgot to update the comment.

varshaprasad96 replied:
> @varshaprasad96 was there a reason you needed the additional script? I would have just added to those tests.

Not exactly - it was to make it easier for review to ensure that apis work for the specific provider as expected.

franciscojavierarceo commented on llama_stack/providers/inline/vector_io/qdrant/config.py:
```diff
@@ -9,15 +9,24 @@
             "path": "${env.QDRANT_PATH:=~/.llama/" + __distro_dir__ + "}/" + "qdrant.db",
+            "kvstore": SqliteKVStoreConfig.sample_run_config(
+                __distro_dir__=__distro_dir__,
+                db_name="qdrant_store.db",
```
nit: by convention I've called all of these <provider/>_registry.db

franciscojavierarceo commented on tests/unit/providers/vector_io/conftest.py:
```diff
@@ -276,16 +278,70 @@ async def chroma_vec_adapter(chroma_vec_db_path, mock_inference_api, embedding_d
 @pytest.fixture
 def vector_io_adapter(vector_provider, request):
     """Returns the appropriate vector IO adapter based on the provider parameter."""
-    vector_provider_dict = {
```
Can you revert back to dictionary? It's cleaner imo

varshaprasad96 commented on llama_stack/providers/utils/memory/vector_store.py:
```diff
@@ -285,6 +285,7 @@ async def insert_chunks(
                 c.embedding = embedding
 
         embeddings = np.array([c.embedding for c in chunks], dtype=np.float32)
+
```
`pre-commit` hook is formatting this.

varshaprasad96 commented on llama_stack/providers/utils/memory/openai_vector_store_mixin.py:
```diff
@@ -259,8 +258,9 @@ async def openai_create_vector_store(
         file_ids = file_ids or []
-        tasks = [self.openai_attach_file_to_vector_store(vector_db_id, file_id) for file_id in file_ids]
-        await asyncio.gather(*tasks)
+        # Process files sequentially to avoid concurrency issues with some vector store providers like qdrant.
+        for file_id in file_ids:
+            await self.openai_attach_file_to_vector_store(vector_db_id, file_id)
```
Qdrant has a limitation wherein `openai_attach_file_to_vector_store` calls `insert_chunks()` - When multiple files are processed concurrently, they all try to write to the same collection simultaneously, which can cause `Storage folder is already accessed by another instance` errors or failed file attachments, making it fragile. This is why I modified the code to process files sequentially, which ensures reliable operation across all vector store providers

ehhuang replied:
This makes *every* provider slower. Can we instead handle the error in `insert_chunks` and retry with backoff if needed?

franciscojavierarceo replied:
+1

varshaprasad96 replied:
Retrying with backoff would still not solve the issue as collection will be locked till it is being accessed for writing a file. I have modified this to be concurrent + falling back to sequential for now.

ehhuang replied:
Thanks for looking into this. Could we please address this at the provider-level (qdrant)? e.g. using a lock on the client.

The reason is that having this makes it silently inefficient for every provider, so I'd prefer to have each provider explicitly address the non-concurrency rather.

varshaprasad96 commented:
> think you accidentally committed two files: `huggingface_datasetio.db` and `localfs_datasetio.db`?

Fixed

varshaprasad96 commented on tests/unit/providers/vector_io/conftest.py:
```diff
@@ -94,7 +96,7 @@ def sample_embeddings_with_metadata(sample_chunks_with_metadata):
 
 
-@pytest.fixture(params=["milvus", "sqlite_vec", "faiss"])
+@pytest.fixture(params=["milvus", "sqlite_vec", "faiss", "chroma", "qdrant"])
```
@franciscojavierarceo I added "chroma" in here - IIRC you were mentioning about this fix today morning.

ehhuang commented on llama_stack/providers/utils/memory/openai_vector_store_mixin.py:
```diff
@@ -264,8 +264,10 @@ async def openai_create_vector_store(
-        tasks = [self.openai_attach_file_to_vector_store(vector_db_id, file_id) for file_id in file_ids]
-        await asyncio.gather(*tasks)
+
+        if file_ids:
```
we can revert this.

ehhuang commented on tests/integration/vector_io/test_openai_vector_stores.py:
```diff
@@ -26,9 +26,26 @@ def skip_if_provider_doesnt_support_openai_vector_stores(client_with_models):
+            "inline::qdrant",
             "remote::pgvector",
-            "remote::chromadb",
+            "remote::qdrant",
```
revert?

ehhuang commented on llama_stack/providers/remote/vector_io/qdrant/qdrant.py:
```diff
@@ -135,17 +146,53 @@ async def delete(self):
+        self.kvstore: KVStore | None = None
+        self.openai_vector_stores: dict[str, dict[str, Any]] = {}
+        # Lock for handling concurrent access to Qdrant storage
+        self._qdrant_lock = asyncio.Lock()
```
I don't see this used?

ehhuang reviewed (commented):
also some tests failing

ehhuang commented on llama_stack/providers/remote/vector_io/qdrant/qdrant.py:
```diff
@@ -135,17 +146,53 @@ async def delete(self):
+        self.vector_db_store = None
+        self.kvstore: KVStore | None = None
+        self.openai_vector_stores: dict[str, dict[str, Any]] = {}
+        # Lock for handling concurrent access to Qdrant storage
```
Please make a comment on how Qdrant doesn't support concurrently adding to index.

ehhuang commented on llama_stack/providers/remote/vector_io/qdrant/qdrant.py:
```diff
@@ -135,17 +146,53 @@ async def delete(self):
-        self.client = AsyncQdrantClient(**self.config.model_dump(exclude_none=True))
+        # Close existing client if it exists
+        # Qdrant doesn't allow multiple clients to access the same storage path simultaneously
+        # This prevents "Storage folder is already accessed by another instance" errors during re-initialization
```
Might be a stupid question: when does re-initialization happen?

varshaprasad96 replied:
That was a comment to be added on `openai_attach_file_to_vector_store`. In tests like https://github.com/meta-llama/llama-stack/blob/406ca72957252409a76fa0ff51b10b8891b07340/tests/unit/providers/vector_io/test_vector_io_openai_vector_stores.py#L64 qdrant would fail without the lock when a new client tries to access the vector store before the previous one is shut down gracefully. It's better to handle this on provider level as other DBs handle them gracefully by default.

varshaprasad96 replied:
There are 2 things which turned our unrelated:
1. Client config initialization - shutting down previous clients before creating a new config (this wasn't needed, so removed it). Thanks for catching it!
2. Concurrent file uploads to the same collection - this turned out to be a known issue in Qdrant - have added comment on this and used provider level lock.

ehhuang reviewed (commented):
A few more comments, otherwise LGTM. Thanks for working on this!

ehhuang commented:
Also please rebase I believe the post-training tests have been disabled.

varshaprasad96 commented:
@ehhuang - Have squashed the Qdrant related changes into a single commit and addressed the comments.

ehhuang reviewed (approved):
Looks great! Thanks for addressing all the comments! Let's make sure CI is green
